src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientFactory.java

```diff
@@ -19,15 +19,29 @@
 
 import com.mongodb.client.MongoClient;
 
+import com.mongodb.spark.sql.connector.annotations.ThreadSafe;
 import com.mongodb.spark.sql.connector.config.Configurable;
 
 /**
  * A factory for creating MongoClients
  *
- * <p><strong>Note:</strong> Care should be taken to implement an {@code MongoClientFactory.equals}
- * method to ensure that the resulting {@link MongoClient} can be cached and reused.
+ * <p>Implementations must have a no-args constructor.
+ *
+ * <p>A factory should produce {@code MongoClients} with the same configuration each time {@link
+ * MongoClientFactory#create()} is called.
+ *
+ * <p>The creation of a {@link MongoClient} is expensive (creating the connection pool and
+ * authenticating connections), so {@code MongoClients} are automatically cached, using the {@code
+ * MongoClientFactory} as the key.
+ *
+ * <p>Implementations of the factory can make use of the automatic caching by ensuring the {@code
+ * MongoClientFactory.equals()} returns true if the {@code MongoClientFactory} instances configured
+ * with the same {@link com.mongodb.spark.sql.connector.config.MongoConfig} and if the
```

Comment:
Both are true. The DefaultMongoClientFactory will produce the same MongoClient per same MongoConfig. It will also produce the same client for any MongoConfig that has the same connection string.

Alternative implementations may choose to match the whole config or parts depending on their usecase.

I will add a clarification.

Reply:
> The DefaultMongoClientFactory will produce the same MongoClient per same MongoConfig.
> It will also produce the same client for any MongoConfig that has the same connection string.

It produces equivalent `MongoClient` objects for equal (not necessary the same) `MongoConfig`s / connection strings. These `MongoClient` objects are neither distinct nor not unequal (`MongoClientImpl` does not override `Object.equals`).

The above fact does not require any changes in the clarified `MongoClientFactory` documentation, but please keep in mind that the terms same/distinct and equal/unequal have very different meaning explained in [JLS 15.21.3](https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.21.3): _"While `==` may be used to compare references of type `String`, such an equality test determines whether or not the two operands refer to the **same** `String` object. The result is `false` if the operands are **distinct** `String` objects, even if they contain the same sequence of characters. The contents of two strings `s` and `t` can be tested for **equality** by the method invocation `s.equals(t)`."_